Content.Server/GameTicking/Rules/GameRuleSystem.Utility.cs

```diff
@@ -88,49 +90,75 @@ protected bool TryGetRandomStation([NotNullWhen(true)] out EntityUid? station, F
             return false;
         }
+        var randomTile = randomTiles.First();
 
-        targetGrid = RobustRandom.Pick(possibleTargets);
-
-        if (!TryComp<MapGridComponent>(targetGrid, out var gridComp))
+        if (!TryComp<MapGridComponent>(randomTile.GridUid, out var gridComp))
+        {
             return false;
+        }
 
-        var found = false;
-        var aabb = gridComp.LocalAABB;
+        tile = randomTile.GridIndices;
+        targetGrid = randomTile.GridUid;
+        targetCoords = _map.GridTileToLocal(targetGrid, gridComp, tile);
 
-        for (var i = 0; i < 10; i++)
-        {
-            var randomX = RobustRandom.Next((int) aabb.Left, (int) aabb.Right);
-            var randomY = RobustRandom.Next((int) aabb.Bottom, (int) aabb.Top);
+        return true;
+    }
+    /// <summary>
+    ///     Get [numberOfTiles] random station tiles, with replacement, that are neither space tiles nor air-blocked tiles
+    /// </summary>
+    protected bool TryFindRandomTilesOnStation(Entity<StationDataComponent> station, int numberOfTiles, out List<TileRef> randomStationTiles)
+    {
+        TryGetAllVaildStationTiles(station, out var allValidStationTiles);
+        var allValidStationTilesCount = allValidStationTiles.Count();
+        randomStationTiles = new();
 
-            tile = new Vector2i(randomX, randomY);
-            if (_atmosphere.IsTileSpace(targetGrid, Transform(targetGrid).MapUid, tile)
-                || _atmosphere.IsTileAirBlocked(targetGrid, tile, mapGridComp: gridComp))
+        if (allValidStationTilesCount == 0)
+        {
+            return false;
+        }
+        for (var i = 0; i < numberOfTiles; i++)
+        {
+            var randomIndex = RobustRandom.Next(allValidStationTilesCount);
+            var curElement = allValidStationTiles.ElementAt(randomIndex);
+            randomStationTiles.Add(curElement);
+        }
+        return true;
+    }
+    /// <summary>
+    ///     Get all station tiles that are neither space tiles nor air-blocked tiles
+    /// </summary>
+    protected bool TryGetAllVaildStationTiles(Entity<StationDataComponent> station, out List<TileRef> allValidStationTiles)
+    {
+        var stationGrids = station.Comp.Grids;
+        allValidStationTiles = new();
+        foreach (var grid in stationGrids)
+        {
+            if (!TryComp<MapGridComponent>(grid, out var gridComp))
             {
                 continue;
             }
-
-            found = true;
-            targetCoords = _map.GridTileToLocal(targetGrid, gridComp, tile);
-            break;
+            foreach (var tile in _map.GetAllTiles(grid, gridComp))
+            {
+                if (tile.IsSpace() || _atmosphere.IsTileAirBlocked(grid, tile.GridIndices, mapGridComp: gridComp))
+                {
+                    continue;
+                }
+                allValidStationTiles.Add(tile);
+            }
         }
-
-        return found;
+        return true;
```

Comment:
1. This method name is wrong.
2. Iterating every single tile up-front is far, far too slow and will just nuke the server for a tick every time this is run.

I've said this before but just pick a random tile and breadth-first search out from it will be far faster and more reliable without being so slow as to tank performance.

Reply:
Wouldn't performing a breadth-first search tend to concentrate messes? Is that not currently the issue?

Reply:
Also, how am I supposed to get that first random tile without searching through the entire grid? From what I can tell, any method that does that is non-public and within RobustToolbox

Reply:
> Wouldn't performing a breadth-first search tend to concentrate messes? Is that not currently the issue?

The issue is 2-fold:
1. It spawns a disproportionate amount on smaller grids.
2. It can spawn in space.

Making the game drop performance is not preferable to what is currently happening. All you need to do is pick a random tile in the AABB to iterate from. Alternatively if you want to make the distribution more even then use fixture AABBs instead.